EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFUnpackerOutput.cc

```diff
@@ -6,72 +6,100 @@
+      else
+	res = static_cast<BMTFCollections*>(coll)->getBMTFMuons();
+
+      //BxBlocks changed the format of the blocks
+      int firstBX = 0, lastBX = 0;
+      int nBX = 0;
+      if (!bxBlocks.empty()) {
+	nBX = bxBlocks[0].header().getTotalBx();//how many BX included in the BxBlock before Suppression
+	getBXRange(nBX, firstBX, lastBX);
+	res->setBXRange(firstBX, lastBX);
+      }
+      else {
+	res->setBXRange(-2, 2);
+	LogDebug("L1T") << "No BXs included in the given Block. Set the BXRange to be (-2,2).";
+	return true;
+      }
+
+      LogDebug("L1T") << "nBX = " << nBX << " firstBX = " << firstBX << " lastBX = " << lastBX;
 			
-			int processor = block.amc().getBoardID() - 1;
-			if ( processor < 0 || processor > 11 )
-			{
-				edm::LogInfo ("l1t:stage2::BMTFUnpackerOutput::unpack") << "Processor found out of range so it will be calculated by the old way";
-				if ( block.amc().getAMCNumber()%2 != 0 )
-			                processor = block.amc().getAMCNumber()/2 ;
-			        else
-			                processor = 6 + (block.amc().getAMCNumber()/2 -1);
-			}
+      int processor = block.amc().getBoardID() - 1;
+      if ( processor < 0 || processor > 11 )
+	{
+	  edm::LogInfo ("l1t:stage2::BMTFUnpackerOutput::unpack") << "Processor found out of range so it will be calculated by the old way";
+	  if ( block.amc().getAMCNumber()%2 != 0 )
+	    processor = block.amc().getAMCNumber()/2 ;
+	  else
+	    processor = 6 + (block.amc().getAMCNumber()/2 -1);
+	}
 
-			for(int ibx = firstBX; ibx <= lastBX; ibx++)
-			{
-				int ip(0);
-				for(unsigned int iw = 0; iw < block.header().getSize()/nBX; iw += 2)
-				{
-					uint32_t raw_first = payload[ip+(ibx+lastBX)*6];
-					ip++;
-					uint32_t raw_secnd = payload[ip+(ibx+lastBX)*6];
-					ip++;
-					if ( raw_first == 0 )
-					{
-						LogDebug("L1T") << "Raw data is zero";
-						continue;
-					}
-					
-					RegionalMuonCand muCand = RegionalMuonCand();
-					RegionalMuonRawDigiTranslator::fillRegionalMuonCand(muCand, raw_first, raw_secnd, processor, tftype::bmtf);
-					muCand.setLink(48 + processor);	//the link corresponds to the uGMT input
+      for (auto bxBlock : bxBlocks) {
+	int ibx = bxBlock.header().getBx();
 
-					LogDebug("L1T") << "Pt = " << muCand.hwPt() << " eta: " << muCand.hwEta() << " phi: " << muCand.hwPhi();
-					if ( muCand.hwQual() != 0 )
-					{
-						res->push_back(ibx, muCand);
-					}
+	for (auto iw = 0; iw < 6; iw+=2) {
+	  uint32_t raw_first = bxBlock.payload()[iw];//payload[ip+(ibx+lastBX)*6];
+	  uint32_t raw_secnd = bxBlock.payload()[iw+1];//payload[ip+(ibx+lastBX)*6];
+	  if ( raw_first == 0 )
+	    {
+	      LogDebug("L1T") << "Raw data is zero";
+	      continue;
+	    }
 					
-				}//for iw
-			}//for ibx
+	  RegionalMuonCand muCand;
+	  RegionalMuonRawDigiTranslator::fillRegionalMuonCand(muCand, raw_first, raw_secnd, processor, tftype::bmtf);
+
+	  if (isKalman) {
+	    //muCand.setLink(????????); to be filled later
```

Comment:
I would set the same link number than for the normal output. If the Kalman will be used at some point the input links at the uGMT will not change.